src/hotspot/share/prims/foreignGlobals.cpp

```diff
@@ -126,62 +126,70 @@ void ArgumentShuffle::print_on(outputStream* os) const {
     to_reg.print_on(os);
     os->print_cr("");
   }
-  os->print_cr("Stack argument bytes: %d", _out_arg_bytes);
   os->print_cr("}");
 }
 
-int NativeCallingConvention::calling_convention(const BasicType* sig_bt, VMStorage* out_regs, int num_args) const {
-  int src_pos = 0;
+int ForeignGlobals::compute_out_arg_bytes(const GrowableArray<VMStorage>& out_regs) {
   uint32_t max_stack_offset = 0;
-  for (int i = 0; i < num_args; i++) {
-    switch (sig_bt[i]) {
-      case T_BOOLEAN:
-      case T_CHAR:
-      case T_BYTE:
-      case T_SHORT:
-      case T_INT:
-      case T_FLOAT: {
-        VMStorage reg = _input_regs.at(src_pos++);
-        out_regs[i] = reg;
-        if (reg.is_stack())
-          max_stack_offset = MAX2(max_stack_offset, reg.offset() + reg.stack_size());
-        break;
-      }
-      case T_LONG:
-      case T_DOUBLE: {
-        assert((i + 1) < num_args && sig_bt[i + 1] == T_VOID, "expecting half");
-        VMStorage reg = _input_regs.at(src_pos++);
-        out_regs[i] = reg;
-        if (reg.is_stack())
-          max_stack_offset = MAX2(max_stack_offset, reg.offset() + reg.stack_size());
-        break;
-      }
-      case T_VOID: // Halves of longs and doubles
-        assert(i != 0 && (sig_bt[i - 1] == T_LONG || sig_bt[i - 1] == T_DOUBLE), "expecting half");
-        out_regs[i] = VMStorage::invalid();
-        break;
-      default:
-        ShouldNotReachHere();
-        break;
-    }
+  for (VMStorage reg : out_regs) {
+    if (reg.is_stack())
+      max_stack_offset = MAX2(max_stack_offset, reg.offset() + reg.stack_size());
   }
   return align_up(max_stack_offset, 8);
 }
 
-int JavaCallingConvention::calling_convention(const BasicType* sig_bt, VMStorage* regs, int num_args) const {
+int ForeignGlobals::java_calling_convention(const BasicType* signature, int num_args, GrowableArray<VMStorage>& out_regs) {
   VMRegPair* vm_regs = NEW_RESOURCE_ARRAY(VMRegPair, num_args);
-  int slots = SharedRuntime::java_calling_convention(sig_bt, vm_regs, num_args);
+  int slots = SharedRuntime::java_calling_convention(signature, vm_regs, num_args);
   for (int i = 0; i < num_args; i++) {
     VMRegPair pair = vm_regs[i];
     // note, we ignore second here. Signature should consist of register-size values. So there should be
     // no need for multi-register pairs.
-    //assert(!pair.first()->is_valid() || pair.is_single_reg(), "must be: %s");
-    regs[i] = as_VMStorage(pair.first(), sig_bt[i]);
+    if (signature[i] != T_VOID) {
+      out_regs.push(as_VMStorage(pair.first(), signature[i]));
+    }
```

Comment:
Since we only care about registers-sized values. It is safe to ignore upper halves of `T_LONG` and `T_DOUBLE` (i.e. `T_VOID`).